### ISSUE TITLE
Remote-Containers Doc: Mention WSL2 specifics in GPG section

### ISSUE THREAD

**Issue by chrmarti** [microsoft/vscode-remote-release#3703]
We could mention the requirement for `socat` in the WSL distro and that a pinentry GUI should be registered on the WSL side.

_Originally posted by @mtsmfm in https://github.com/microsoft/vscode-remote-release/issues/3168#issuecomment-695762133_:

https://github.com/microsoft/vscode-remote-release/issues/3168#issuecomment-674355155

Finally I've found the solution for my problem (`gpg: signing failed: Not a tty`).

My environment:

- [Host OS] Windows 10
- [WSL2 OS] Ubuntu
- Dev Container is opened from WSL2

1. [Windows] Install Gpg4win
  I just need pinentry.exe, probably we can install more suitable one on WSL2
2. [WSL2] Install socat (`sudo apt install socat`)
  TBH, I'm not sure why it's needed though
3. [WSL2] Install gpg (`sudo apt install gpg`)
    > In that case we connect the container's GPG to WSL's GPG socket
    https://github.com/microsoft/vscode-remote-release/issues/3168#issuecomment-689313475

    I need to setup gpg stuff on WSL2, not on Windows. It's really confusing 😕 
3. [WSL2] Set pinentry-program option ` echo pinentry-program /mnt/c/Program\ Files\ \(x86\)/Gpg4win/bin/pinentry.exe > ~/.gnupg/gpg-agent.conf`
4. [WSL2] Reload gpg agent on WSL2 ` gpg-connect-agent reloadagent /bye`
5. Now I can run `gpg --output test.sig --detach-sig test.txt`

**Comment by exhuma** [comment]
Following the steps here I was able to sign files inside WSL2, but *not* inside a dev-container.

As I understand the comments, the above steps are mainly used to use the pinentry-program from GPG4win (nothing else of it is actually used).

In my first try I loaded my private key into the GPG4Win agent (kleopatra if I'm not mistaken) but noticed it wasn't working (see also #3702). Upon rereading the steps above, this line pointed me to actually install gpg and import my key *inside WSL2*!

> I need to setup gpg stuff on WSL2, not on Windows. It's really confusing 😕

After doing so, gpg works inside WSL2. But I am still unable to sign commits inside a dev-container.

**Question:** Is the solution by @mtsmfm supposed to allow signing inside WSL2 or inside the dev-container? If it's the latter, what am I missing to make this work?

**Comment by mtsmfm** [comment]
> Question: Is the solution by @mtsmfm supposed to allow signing inside WSL2 or inside the dev-container? If it's the latter, what am I missing to make this work?

@exhuma Both. 
However for some reason sometimes invoking `pinentry-program` fails because WSL fails to communicate with Windows.
I need to restart my PC in such a case.

**Comment by exhuma** [comment]
Without doing any changes concerning `gpg` it... "magically" started working now. I have rebuilt the dev-container for other reasons though. So maybe rebuilding the dev-container did something useful here. So my steps were - roughly - as follows (in the hopes it helps someone else):

* Set up a dev-container with dev-volume
* Try committing changes and notice GPG is not working
* Follow the steps outlined in the above post
* Notice GPG is still not working
* Reach out on GitHub in this thread for help and continue normal development without GPG signing, waiting for feedback
* By chance, the changes I made were in the Dockerfile and I needed to rebuild the dev-container
* After lots of hacking and rebuilding (without testing gpg-signing during that time) I *finally* had something commit-worthy and noticed the signing suddenly worked.

I *believe* that the container-rebuild might be responsible for the "fix". But I also made changes in my dotfiles during that process, although these changes should be unrelated.
